Introduce a bulk format that uses a prefix length

iverase opened the pull request:
The bulk API currently uses a suffix marker to signal the end of a document. This way of breaking the documents has a few drawbacks:

1) It requires to parse all the bytes of a document in order to find the mark 
2) It s only properly supported for Json content type. It can in theory be used for Smile format too but it breaks as the [document might contain the marker](https://github.com/elastic/elasticsearch/issues/94319)

This change proposes to add an alternative format for the bulk API that uses a 4 byte prefix for each document containing the length of the document in big endian order.  This makes the bulk easier to parse as we don't need to read every byte to know the length of the document and it supports any type of content type.

fixes https://github.com/elastic/elasticsearch/issues/94319

iverase committed:
[Draft] Introduce a bulk format that uses a prefix length

iverase committed:
iter

elasticsearchmachine commented:
Hi @iverase, I've created a changelog YAML for you.

iverase committed:
Update docs/changelog/135506.yaml

iverase committed:
iter

iverase committed:
Merge branch 'prefix-length-bulk' of github.com:iverase/elasticsearch into prefix-length-bulk

iverase committed:
iter

ChrisHegarty commented on server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java:
```diff
@@ -275,27 +295,59 @@ private int tryParse(BytesReference data, boolean lastData) throws IOException {
+                            "Documents in the bulk request must be prefixed with the length of the document"
+                        );
+                    }
+                    break;
```
I'm unsure of the exact control flow here so I could be wrong, but if we were to break here should `from` already be updated? since we read the bytes from `data`? Or is it deliberate to effectively leave the 4 byte length as unconsumed?

iverase committed:
improve test

iverase committed:
Merge branch 'main' into prefix-length-bulk

iverase committed:
Integration tests

iverase committed:
Update docs/changelog/135506.yaml

iverase committed:
Merge branch 'main' into prefix-length-bulk

benwtrent reviewed (commented):
I really like how simple this is! I am all for it :)

Some comments on the header name, and looks like some left over debugging stuff?

iverase committed:
Merge branch 'main' into prefix-length-bulk

iverase committed:
iter

iverase committed:
Merge branch 'prefix-length-bulk' of github.com:iverase/elasticsearch into prefix-length-bulk

iverase commented:
I wan to to point out that the only formats currently accepted are JSON and SMILE as the RestController has code to reject the request otherwise:

https://github.com/elastic/elasticsearch/blob/eeca493860be1936ce099533aaa53d481ca73071/server/src/main/java/org/elasticsearch/rest/RestController.java#L426

We can in theory support any format with this change so we might need to change this if we want to support them.

iverase committed:
Merge branch 'main' into prefix-length-bulk

DaveCTurner commented on server/src/main/java/org/elasticsearch/rest/action/document/RestBulkAction.java:
```diff
@@ -62,6 +64,29 @@ public class RestBulkAction extends BaseRestHandler {
     public static final String FAILURE_STORE_STATUS_CAPABILITY = "failure_store_status";
+    public static final String PREFIX_LENGTH_FORMAT_CAPABILITY = "prefix_length_format";
+
+    private static final String BULK_FORMAT_HEADER = "X-Bulk-Format";
```
I do not think we should introduce a separate header for this feature. Instead, we should let clients say that the request body is a length-prefixed bulk request using an appropriate value for the existing `Content-type` header. We already define a few vendor-specific content types with the `application/vnd.elasticsearch+` prefix. Let's come up with a way to add some more of them.

DaveCTurner replied:
Yes that's my thinking.

I'd suggest we stick with the `application/vnd.elasticsearch+` prefix since this is a vendor-specific format we're defining. I also don't think we need to fiddle with a separate `format=` parameter, we can leave `application/x-ndjson` as (effectively) the only delimited format and _define_ `application/vnd.elasticsearch+json.stream` etc as being length-prefixed.

> for json I think we should stick with ascii encoding

FWIW I disagree, I'd rather we used the same length-encoding scheme for all formats, and I'd rather it was a fixed number of bytes. Length-prefixed messages are awful to construct in a text editor regardless of the length encoding, you'd be kinda mad not to use some simple script to do this instead.

iverase replied:
Having Content-Type headers that do not resolve on XContentType adds an incredible amount of complexity in how rest request are handled. Not sure how to proceed, do you have any suggestion?

iverase replied:
First try by adding a XContentLengthPrefixedStreamingType in https://github.com/elastic/elasticsearch/pull/135506/commits/fc7513aaf1b291a1140615f6b08f626786075b16

DaveCTurner reviewed (commented):
This is a great idea, thanks for tackling it. I have one concern, see inline comment.

iverase committed:
Add XContentLengthPrefixedStreamingType

iverase committed:
Merge branch 'main' into prefix-length-bulk

iverase committed:
iter

iverase committed:
iter

iverase committed:
renames

iverase committed:
iter

iverase committed:
Merge branch 'main' into prefix-length-bulk

DaveCTurner commented:
Overall seems reasonable to me, sorry I didn't realise this was blocked awaiting further comments.

I don't think they need to return a `queryParameter()` - this only relates to the formatting of a response, but we won't (yet) be supporting `?format=stream.cbor` etc to specify the response format. I'd prefer we adjusted `MediaTypeRegistry` to accept a `null` response from this method for request-only media types.

I'm a little against further overloading `org.elasticsearch.rest.RestRequest#xContentType` to represent both the type of the body of a traditional one-shot request and the type of the items in a new streaming request. It's already overloaded like this for existing bulk requests but this is making it slightly worse. Not a blocker tho, we can tidy this up later if needed. A code comment about the exact meaning of this field would be appreciated.

iverase committed:
address review comments

iverase commented:
Thanks @DaveCTurner , I addressed your comments. Now `MediaTypeRegistry` accepts null as query parameter and `.RestRequest#xContentType` will be null if content-type  contains a streaming media type. The result of the bulk will be json except when the user request a different output via the `Accept` header.

iverase committed:
Merge branch 'main' into prefix-length-bulk

iverase committed:
Merge branch 'main' into prefix-length-bulk

henningandersen reviewed (commented):
I wonder if we need all the new variants or if we could make do with just one of them? I assume one of the binary ones is the primary target.

I am not sure the length-encoding is my preference or if we should see it more as an array of cbor/smile. That would fit other binary formats like a binary csv representation better. I suppose we can have both, but it is not clear to me we need all that variety of format support.

iverase committed:
Merge branch 'main' into prefix-length-bulk

iverase committed:
iter

DaveCTurner commented:
> or if we should see it more as an array of cbor/smile

This would require the coordinating node to parse the document (not necessarily creating any objects) just to find the object boundaries. I expect that'd perform worse than looking for newlines in NDJSON like we do today. The advantage of the length prefix is that the coordinating node can (in principle at least) treat the documents as opaque sequences of bytes that get forwarded on to the relevant data nodes. We can't quite get it down to zero-copy because of TLS but we could get pretty close.
